qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -314,6 +311,15 @@ def _reset_qubits_decay(self):
         """
         self.qubits_decay = {k: 1 for k in self.qubits_decay.keys()}
 
+    def _build_required_predecessors(self, dag):
+        out = defaultdict(int)
+        # We don't need to count in- or out-wires: outs can never be predecessors, and all input
+        # wires are automatically satisfied at the start.
+        for node in dag.op_nodes():
+            for successor in self._successors(node, dag):
+                out[successor] += 1
```

Comment:
I'm wondering if doing something like:
```suggestion
        for node in dag.op_nodes():
            out_successor[node] += dag._multi_graph.in_degree(node)
```

would work and be faster. Not that I expect this to be a bottleneck. It just took me a sec to grok this.
I guess there might be an issue because of the input nodes as we'd be including the first layer nodes and they'd have a required predecessor count for the inputs. That's not too different than what this outputs now because we'll end up with output nodes in the `out` dict but that should be ok because we'll never do a lookup on an output wire.

Reply:
I think this will count `DAGInNodes` too?  We ought to discount those to do things properly, but we could just loop through the original `front_layer` and automatically set those to zero.

Reply:
Oh, I clearly didn't read your comment properly sorry!

Reply:
Lets just leave this as is. We can revisit this after #7977 and see if we're actually seeing this in profiles or not.